In [83]:
import pandas as pd
import csv

In [3]:
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)

<hr>

# Output from Problem 1: we have filename and calculated distance

In [4]:
output = pd.read_csv("output.csv")

In [5]:
output.head()

filename   distance
0  9598683_040_pred.png  23.953488
1  9063955_094_pred.png        NaN
2  9579246_002_pred.png        NaN
3  9802170_058_pred.png  22.187050
4  9950790_056_pred.png  18.023077

In [6]:
output['filename'] = output['filename'].astype(str)
output["filename"] = output["filename"].str.replace("_pred.png", "", regex=False)

In [7]:
output.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31680 entries, 0 to 31679
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   filename  31680 non-null  object 
 1   distance  17012 non-null  float64
dtypes: float64(1), object(1)
memory usage: 495.1+ KB


In [8]:
output = output.sort_values(by="filename").reset_index(drop=True)

In [9]:
output[["case", "slice"]] = output["filename"].str.split("_", expand=True)

In [10]:
output = output[['filename', 'case', 'slice', 'distance']]

In [11]:
# Example of case 9002116 with label B (OA: 3 to 4)
output.head(160)

filename     case slice   distance
0    9002116_001  9002116   001        NaN
1    9002116_002  9002116   002        NaN
2    9002116_003  9002116   003        NaN
3    9002116_004  9002116   004        NaN
4    9002116_005  9002116   005        NaN
5    9002116_006  9002116   006        NaN
6    9002116_007  9002116   007        NaN
7    9002116_008  9002116   008        NaN
8    9002116_009  9002116   009        NaN
9    9002116_010  9002116   010        NaN
10   9002116_011  9002116   011        NaN
11   9002116_012  9002116   012        NaN
12   9002116_013  9002116   013        NaN
13   9002116_014  9002116   014        NaN
14   9002116_015  9002116   015        NaN
15   9002116_016  9002116   016        NaN
16   9002116_017  9002116   017        NaN
17   9002116_018  9002116   018        NaN
18   9002116_019  9002116   019        NaN
19   9002116_020  9002116   020        NaN
20   9002116_021  9002116   021        NaN
21   9002116_022  9002116   022        NaN
22   9002116_023  9002116   023        NaN
23   9002116_024  9002116   024        NaN
24   9002116_025  9002116   025        NaN
25   9002116_026  9002116   026        NaN
26   9002116_027  9002116   027  17.895522
27   9002116_028  9002116   028  15.802632
28   9002116_029  9002116   029  14.926829
29   9002116_030  9002116   030  14.022727
30   9002116_031  9002116   031  13.265957
31   9002116_032  9002116   032  13.181818
32   9002116_033  9002116   033  12.754902
33   9002116_034  9002116   034  12.745283
34   9002116_035  9002116   035  13.036364
35   9002116_036  9002116   036  13.434783
36   9002116_037  9002116   037  13.372881
37   9002116_038  9002116   038  13.341667
38   9002116_039  9002116   039  13.090164
39   9002116_040  9002116   040  12.975806
40   9002116_041  9002116   041  13.070312
41   9002116_042  9002116   042  13.213740
42   9002116_043  9002116   043  13.146154
43   9002116_044  9002116   044  13.083969
44   9002116_045  9002116   045  12.932331
45   9002116_046  9002116   046  13.118519
46   9002116_047  9002116   047  13.389706
47   9002116_048  9002116   048  13.877698
48   9002116_049  9002116   049  14.433566
49   9002116_050  9002116   050  14.972414
50   9002116_051  9002116   051  15.074830
51   9002116_052  9002116   052  15.229730
52   9002116_053  9002116   053  15.553333
53   9002116_054  9002116   054  15.921569
54   9002116_055  9002116   055  16.311688
55   9002116_056  9002116   056  16.167742
56   9002116_057  9002116   057  16.547771
57   9002116_058  9002116   058  16.905660
58   9002116_059  9002116   059  17.525000
59   9002116_060  9002116   060  17.616352
60   9002116_061  9002116   061  17.394904
61   9002116_062  9002116   062  17.583333
62   9002116_063  9002116   063  18.382166
63   9002116_064  9002116   064  18.923077
64   9002116_065  9002116   065  19.743590
65   9002116_066  9002116   066  20.655844
66   9002116_067  9002116   067  21.228758
67   9002116_068  9002116   068  21.947368
68   9002116_069  9002116   069  23.420000
69   9002116_070  9002116   070  23.310345
70   9002116_071  9002116   071  23.281690
71   9002116_072  9002116   072  20.208333
72   9002116_073  9002116   073  28.365217
73   9002116_074  9002116   074  29.808696
74   9002116_075  9002116   075  30.379310
75   9002116_076  9002116   076        NaN
76   9002116_077  9002116   077        NaN
77   9002116_078  9002116   078        NaN
78   9002116_079  9002116   079        NaN
79   9002116_080  9002116   080        NaN
80   9002116_081  9002116   081        NaN
81   9002116_082  9002116   082        NaN
82   9002116_083  9002116   083        NaN
83   9002116_084  9002116   084        NaN
84   9002116_085  9002116   085        NaN
85   9002116_086  9002116   086        NaN
86   9002116_087  9002116   087        NaN
87   9002116_088  9002116   088        NaN
88   9002116_089  9002116   089        NaN
89   9002116_090  9002116   090        NaN
90   9002116_091  9002116   091        NaN
91   9002116_092  9002116   092        NaN
92   9002

In [12]:
# Example of case 9005075 with label A (non-OA: 0 to 2)
output.iloc[160:320]

filename     case slice   distance
160  9005075_001  9005075   001        NaN
161  9005075_002  9005075   002        NaN
162  9005075_003  9005075   003        NaN
163  9005075_004  9005075   004        NaN
164  9005075_005  9005075   005        NaN
165  9005075_006  9005075   006        NaN
166  9005075_007  9005075   007        NaN
167  9005075_008  9005075   008        NaN
168  9005075_009  9005075   009        NaN
169  9005075_010  9005075   010        NaN
170  9005075_011  9005075   011        NaN
171  9005075_012  9005075   012        NaN
172  9005075_013  9005075   013        NaN
173  9005075_014  9005075   014        NaN
174  9005075_015  9005075   015        NaN
175  9005075_016  9005075   016        NaN
176  9005075_017  9005075   017        NaN
177  9005075_018  9005075   018        NaN
178  9005075_019  9005075   019        NaN
179  9005075_020  9005075   020        NaN
180  9005075_021  9005075   021        NaN
181  9005075_022  9005075   022        NaN
182  9005075_023  9005075   023        NaN
183  9005075_024  9005075   024        NaN
184  9005075_025  9005075   025        NaN
185  9005075_026  9005075   026        NaN
186  9005075_027  9005075   027  72.307692
187  9005075_028  9005075   028  65.549020
188  9005075_029  9005075   029  61.066667
189  9005075_030  9005075   030  31.630952
190  9005075_031  9005075   031  31.597826
191  9005075_032  9005075   032  29.600000
192  9005075_033  9005075   033  27.888889
193  9005075_034  9005075   034  26.766990
194  9005075_035  9005075   035  25.409524
195  9005075_036  9005075   036  24.179245
196  9005075_037  9005075   037  23.289720
197  9005075_038  9005075   038  23.146789
198  9005075_039  9005075   039  22.554545
199  9005075_040  9005075   040  22.522124
200  9005075_041  9005075   041  22.184211
201  9005075_042  9005075   042  21.913043
202  9005075_043  9005075   043  21.530435
203  9005075_044  9005075   044  23.117647
204  9005075_045  9005075   045  22.663866
205  9005075_046  9005075   046  22.050847
206  9005075_047  9005075   047  22.403361
207  9005075_048  9005075   048  22.766667
208  9005075_049  9005075   049  23.557377
209  9005075_050  9005075   050  23.747967
210  9005075_051  9005075   051  24.370968
211  9005075_052  9005075   052  25.008000
212  9005075_053  9005075   053  27.148438
213  9005075_054  9005075   054  29.389313
214  9005075_055  9005075   055  29.887218
215  9005075_056  9005075   056  32.569343
216  9005075_057  9005075   057  33.394161
217  9005075_058  9005075   058  34.316547
218  9005075_059  9005075   059  35.964539
219  9005075_060  9005075   060  36.773050
220  9005075_061  9005075   061  36.237410
221  9005075_062  9005075   062  38.524823
222  9005075_063  9005075   063  38.414286
223  9005075_064  9005075   064  38.690647
224  9005075_065  9005075   065  39.597122
225  9005075_066  9005075   066  40.304348
226  9005075_067  9005075   067  40.802920
227  9005075_068  9005075   068  40.778626
228  9005075_069  9005075   069  41.892308
229  9005075_070  9005075   070        NaN
230  9005075_071  9005075   071        NaN
231  9005075_072  9005075   072        NaN
232  9005075_073  9005075   073        NaN
233  9005075_074  9005075   074        NaN
234  9005075_075  9005075   075        NaN
235  9005075_076  9005075   076        NaN
236  9005075_077  9005075   077        NaN
237  9005075_078  9005075   078        NaN
238  9005075_079  9005075   079        NaN
239  9005075_080  9005075   080        NaN
240  9005075_081  9005075   081        NaN
241  9005075_082  9005075   082        NaN
242  9005075_083  9005075   083        NaN
243  9005075_084  9005075   084        NaN
244  9005075_085  9005075   085        NaN
245  9005075_086  9005075   086        NaN
246  9005075_087  9005075   087        NaN
247  9005075_088  9005075   088        NaN
248  9005075_089  9005075   089        NaN
249  9005075_090  9005075   090        NaN
250  9005075_091  9005075   091        NaN
251  9005075_092  9005075   092        NaN
252  9005

In [13]:
output.tail(160)

filename     case slice   distance
31520  9997856_001  9997856   001        NaN
31521  9997856_002  9997856   002        NaN
31522  9997856_003  9997856   003        NaN
31523  9997856_004  9997856   004        NaN
31524  9997856_005  9997856   005        NaN
31525  9997856_006  9997856   006        NaN
31526  9997856_007  9997856   007        NaN
31527  9997856_008  9997856   008        NaN
31528  9997856_009  9997856   009        NaN
31529  9997856_010  9997856   010        NaN
31530  9997856_011  9997856   011        NaN
31531  9997856_012  9997856   012        NaN
31532  9997856_013  9997856   013        NaN
31533  9997856_014  9997856   014        NaN
31534  9997856_015  9997856   015        NaN
31535  9997856_016  9997856   016        NaN
31536  9997856_017  9997856   017        NaN
31537  9997856_018  9997856   018        NaN
31538  9997856_019  9997856   019        NaN
31539  9997856_020  9997856   020  61.441176
31540  9997856_021  9997856   021  58.783784
31541  9997856_022  9997856   022  32.121951
31542  9997856_023  9997856   023        NaN
31543  9997856_024  9997856   024  45.019231
31544  9997856_025  9997856   025  23.083333
31545  9997856_026  9997856   026  18.069444
31546  9997856_027  9997856   027  17.236842
31547  9997856_028  9997856   028  16.746988
31548  9997856_029  9997856   029  17.375000
31549  9997856_030  9997856   030  18.297872
31550  9997856_031  9997856   031  18.696970
31551  9997856_032  9997856   032  17.500000
31552  9997856_033  9997856   033  17.429907
31553  9997856_034  9997856   034  17.189189
31554  9997856_035  9997856   035  17.226087
31555  9997856_036  9997856   036  17.302521
31556  9997856_037  9997856   037  17.147541
31557  9997856_038  9997856   038  17.388889
31558  9997856_039  9997856   039  17.061538
31559  9997856_040  9997856   040  16.857143
31560  9997856_041  9997856   041  16.941176
31561  9997856_042  9997856   042  16.942857
31562  9997856_043  9997856   043  16.542254
31563  9997856_044  9997856   044  16.455172
31564  9997856_045  9997856   045  16.197279
31565  9997856_046  9997856   046  16.306667
31566  9997856_047  9997856   047  16.274510
31567  9997856_048  9997856   048  16.141935
31568  9997856_049  9997856   049  16.108280
31569  9997856_050  9997856   050  16.275000
31570  9997856_051  9997856   051  15.801242
31571  9997856_052  9997856   052  15.635802
31572  9997856_053  9997856   053  15.456790
31573  9997856_054  9997856   054  15.581818
31574  9997856_055  9997856   055  15.432927
31575  9997856_056  9997856   056  15.716867
31576  9997856_057  9997856   057  17.187135
31577  9997856_058  9997856   058  18.097701
31578  9997856_059  9997856   059  18.202312
31579  9997856_060  9997856   060  18.802326
31580  9997856_061  9997856   061  24.571429
31581  9997856_062  9997856   062  23.660000
31582  9997856_063  9997856   063  24.891892
31583  9997856_064  9997856   064  27.571429
31584  9997856_065  9997856   065  25.651852
31585  9997856_066  9997856   066  30.000000
31586  9997856_067  9997856   067  29.719424
31587  9997856_068  9997856   068  31.892857
31588  9997856_069  9997856   069  32.287770
31589  9997856_070  9997856   070  33.413043
31590  9997856_071  9997856   071  34.355556
31591  9997856_072  9997856   072        NaN
31592  9997856_073  9997856   073  37.413534
31593  9997856_074  9997856   074  36.545455
31594  9997856_075  9997856   075  35.796875
31595  9997856_076  9997856   076  37.692308
31596  9997856_077  9997856   077  38.830769
31597  9997856_078  9997856   078  39.830769
31598  9997856_079  9997856   079  41.473282
31599  9997856_080  9997856   080  43.115385
31600  9997856_081  9997856   081  44.863636
31601  9997856_082  9997856   082  47.090226
31602  9997856_083  9997856   083  47.611940
31603  9997856_084  9997856   084  47.259259
31604  9997856_085  9997856   085  41.169118
31605  9997856_086  9997856   086  40.832117
31606  9997856_087  9997856   087  38.231788
31607  9997856_088  9997856   088  38.401316
31608

In [14]:
output['case'].nunique()

198

In [15]:
case_list = output['case'].unique().tolist()

In [16]:
type(case_list[0])

str

In [17]:
case_list.sort()

In [18]:
len(case_list)

198

In [19]:
case_list

['9002116',
 '9005075',
 '9005132',
 '9026934',
 '9030718',
 '9031141',
 '9033275',
 '9037952',
 '9043945',
 '9047800',
 '9048789',
 '9049007',
 '9058960',
 '9063955',
 '9064631',
 '9065272',
 '9072646',
 '9074878',
 '9081858',
 '9086868',
 '9088414',
 '9089119',
 '9101270',
 '9123404',
 '9126260',
 '9135752',
 '9138965',
 '9155861',
 '9157453',
 '9159804',
 '9167541',
 '9173013',
 '9188806',
 '9189986',
 '9208016',
 '9238970',
 '9253530',
 '9270239',
 '9273764',
 '9282257',
 '9291078',
 '9294761',
 '9300338',
 '9310416',
 '9326657',
 '9328093',
 '9330729',
 '9341240',
 '9369225',
 '9372474',
 '9373191',
 '9374235',
 '9386206',
 '9393987',
 '9402139',
 '9406962',
 '9410253',
 '9414291',
 '9414391',
 '9427858',
 '9430102',
 '9437407',
 '9438742',
 '9438852',
 '9441420',
 '9448098',
 '9452100',
 '9458093',
 '9458176',
 '9461404',
 '9467278',
 '9467311',
 '9469520',
 '9472254',
 '9475264',
 '9488441',
 '9493073',
 '9493245',
 '9493338',
 '9494867',
 '9495437',
 '9495873',
 '9497375',
 '95

<hr>

In [20]:
data = pd.read_csv("../KL grade for baseline year and 2-year follow up.csv")

In [21]:
data['cases'] = data['cases'].astype(str)

In [22]:
data.head()

cases  V00XRKL V03KL
0  9002116        3     B
1  9005075        0     A
2  9005132        4     B
3  9026934        2     A
4  9030718        3     B

In [23]:
data['cases'].nunique()

196

In [24]:
case_list_2 = data['cases'].unique().tolist()

In [25]:
case_list_2.sort()

In [26]:
type(case_list_2[0])

str

In [27]:
case_list_2

['9002116',
 '9005075',
 '9005132',
 '9026934',
 '9030718',
 '9031141',
 '9033275',
 '9037952',
 '9043945',
 '9047800',
 '9048789',
 '9049007',
 '9058960',
 '9063955',
 '9064631',
 '9065272',
 '9072646',
 '9074878',
 '9081858',
 '9086868',
 '9088414',
 '9089119',
 '9101270',
 '9123404',
 '9126260',
 '9135752',
 '9138965',
 '9155861',
 '9157453',
 '9159804',
 '9167541',
 '9173013',
 '9188806',
 '9189986',
 '9208016',
 '9238970',
 '9253530',
 '9270239',
 '9273764',
 '9282257',
 '9291078',
 '9294761',
 '9300338',
 '9310416',
 '9326657',
 '9328093',
 '9330729',
 '9341240',
 '9369225',
 '9372474',
 '9373191',
 '9374235',
 '9386206',
 '9393987',
 '9402139',
 '9406962',
 '9410253',
 '9414291',
 '9414391',
 '9427858',
 '9430102',
 '9437407',
 '9438742',
 '9438852',
 '9441420',
 '9448098',
 '9452100',
 '9458093',
 '9458176',
 '9461404',
 '9467278',
 '9467311',
 '9469520',
 '9472254',
 '9475264',
 '9488441',
 '9493073',
 '9493245',
 '9493338',
 '9494867',
 '9495437',
 '9495873',
 '9497375',
 '95

In [28]:
if case_list == case_list_2:
    print("The case lists are identical.")
else:
    diff1 = set(case_list) - set(case_list_2)
    diff2 = set(case_list_2) - set(case_list)

    print("Items in case_list but not in case_list_2:", diff1)
    print("Items in case_list_2 but not in case_list:", diff2)

Items in case_list but not in case_list_2: {'9566972', '9633944'}
Items in case_list_2 but not in case_list: set()


* There are 2 cases in 'preds' folder that are not in "KL grade for baseline year and 2-year follow up.csv"

* These 2 cases don't have KL grades and labels (A/B)

* Will be removed.

In [29]:
cases_to_removed = output[output['case'].isin(diff1)]

In [30]:
cases_to_removed.shape

(320, 4)

* Drop 2 cases {'9566972', '9633944'} : 160 slices x 2 = 320 rows

In [31]:
output.shape

(31680, 4)

In [32]:
output = output.drop(cases_to_removed.index).reset_index(drop=True)

In [33]:
output.shape

(31360, 4)

In [34]:
output['slice'] = output['slice'].astype(int)

In [35]:
output.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31360 entries, 0 to 31359
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   filename  31360 non-null  object 
 1   case      31360 non-null  object 
 2   slice     31360 non-null  int64  
 3   distance  16881 non-null  float64
dtypes: float64(1), int64(1), object(2)
memory usage: 980.1+ KB


In [36]:
output.head()

filename     case  slice  distance
0  9002116_001  9002116      1       NaN
1  9002116_002  9002116      2       NaN
2  9002116_003  9002116      3       NaN
3  9002116_004  9002116      4       NaN
4  9002116_005  9002116      5       NaN

<hr>

# Output from Problem 2: we have start and end range of slices for each case

In [37]:
ranges_df = pd.read_csv("valid_slice_ranges.csv")

In [38]:
ranges_df['case_id'] = ranges_df['case_id'].astype(str)

In [39]:
ranges_df.head()

case_id  start_slice  end_slice  num_valid_slices
0  9002116           33        140               106
1  9005075           30        136               107
2  9005132           30        139               107
3  9026934           38        129                92
4  9030718           24        114                91

In [40]:
ranges_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 198 entries, 0 to 197
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   case_id           198 non-null    object
 1   start_slice       198 non-null    int64 
 2   end_slice         198 non-null    int64 
 3   num_valid_slices  198 non-null    int64 
dtypes: int64(3), object(1)
memory usage: 6.3+ KB


In [41]:
ranges_df['case_id'].nunique()

198

In [42]:
{'9566972', '9633944'}
ranges_df[ranges_df['case_id'].isin({'9566972', '9633944'})]

case_id  start_slice  end_slice  num_valid_slices
97   9566972           27        110                77
113  9633944           41        128                88

In [43]:
ranges_df = ranges_df[~ranges_df['case_id'].isin({'9566972', '9633944'})].reset_index(drop=True)

In [44]:
ranges_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 196 entries, 0 to 195
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   case_id           196 non-null    object
 1   start_slice       196 non-null    int64 
 2   end_slice         196 non-null    int64 
 3   num_valid_slices  196 non-null    int64 
dtypes: int64(3), object(1)
memory usage: 6.3+ KB


In [45]:
ranges_df = ranges_df.sort_values(by="case_id").reset_index(drop=True)

In [46]:
new_case_list = ranges_df['case_id'].unique().tolist()

In [47]:
type(new_case_list[0])

str

In [48]:
new_case_list.sort()

In [49]:
len(new_case_list)

196

In [50]:
new_case_list == case_list

False

In [51]:
new_case_list == case_list_2

True

<hr>

In [52]:
# create a new df to store the filtered slices:

df = pd.DataFrame(columns=["case", "slice", "distance"])

In [53]:
for case, slice, distance in zip(output['case'], output['slice'], output['distance']):
    for case_id, start, end in zip(ranges_df['case_id'], ranges_df['start_slice'], ranges_df['end_slice']):
        if case == case_id and start <= slice <= end:
            df.loc[len(df)] = {
                "case": case,
                "slice": slice,
                "distance": distance
            }

In [54]:
df.head()

case  slice   distance
0  9002116     33  12.754902
1  9002116     34  12.745283
2  9002116     35  13.036364
3  9002116     36  13.434783
4  9002116     37  13.372881

<hr>

# Verify if the slice is selected within range

In [55]:
ranges_df[ranges_df['case_id'] == '9002116']

case_id  start_slice  end_slice  num_valid_slices
0  9002116           33        140               106

In [56]:
df[df['case'] == '9002116']

case  slice   distance
0    9002116     33  12.754902
1    9002116     34  12.745283
2    9002116     35  13.036364
3    9002116     36  13.434783
4    9002116     37  13.372881
5    9002116     38  13.341667
6    9002116     39  13.090164
7    9002116     40  12.975806
8    9002116     41  13.070312
9    9002116     42  13.213740
10   9002116     43  13.146154
11   9002116     44  13.083969
12   9002116     45  12.932331
13   9002116     46  13.118519
14   9002116     47  13.389706
15   9002116     48  13.877698
16   9002116     49  14.433566
17   9002116     50  14.972414
18   9002116     51  15.074830
19   9002116     52  15.229730
20   9002116     53  15.553333
21   9002116     54  15.921569
22   9002116     55  16.311688
23   9002116     56  16.167742
24   9002116     57  16.547771
25   9002116     58  16.905660
26   9002116     59  17.525000
27   9002116     60  17.616352
28   9002116     61  17.394904
29   9002116     62  17.583333
30   9002116     63  18.382166
31   9002116     64  18.923077
32   9002116     65  19.743590
33   9002116     66  20.655844
34   9002116     67  21.228758
35   9002116     68  21.947368
36   9002116     69  23.420000
37   9002116     70  23.310345
38   9002116     71  23.281690
39   9002116     72  20.208333
40   9002116     73  28.365217
41   9002116     74  29.808696
42   9002116     75  30.379310
43   9002116     76        NaN
44   9002116     77        NaN
45   9002116     78        NaN
46   9002116     79        NaN
47   9002116     80        NaN
48   9002116     81        NaN
49   9002116     82        NaN
50   9002116     83        NaN
51   9002116     84        NaN
52   9002116     85        NaN
53   9002116     86        NaN
54   9002116     87        NaN
55   9002116     88        NaN
56   9002116     89        NaN
57   9002116     90        NaN
58   9002116     91        NaN
59   9002116     92        NaN
60   9002116     93  46.592000
61   9002116     94  45.257812
62   9002116     95  42.435115
63   9002116     96  43.089172
64   9002116     97  42.745342
65   9002116     98  41.527950
66   9002116     99  41.962963
67   9002116    100  41.672840
68   9002116    101  41.441718
69   9002116    102  40.765432
70   9002116    103  42.236364
71   9002116    104  40.274390
72   9002116    105  40.192771
73   9002116    106  39.724551
74   9002116    107  38.748503
75   9002116    108  38.023810
76   9002116    109  37.244048
77   9002116    110  36.633136
78   9002116    111  35.686391
79   9002116    112  34.221557
80   9002116    113  34.303571
81   9002116    114  32.243902
82   9002116    115  32.036810
83   9002116    116  30.925000
84   9002116    117  29.318471
85   9002116    118  28.051948
86   9002116    119  24.027211
87   9002116    120  24.231293
88   9002116    121  24.205479
89   9002116    122  23.924658
90   9002116    123  24.260274
91   9002116    124  23.524138
92   9002116    125  23.160839
93   9002116    126  22.895833
94   9002116    127  23.097902
95   9002116    128  23.244755
96   9002116    129  23.316901
97   9002116    130  24.517241
98   9002116    131  24.447552
99   9002116    132  25.042553
100  9002116    133  25.935252
101  9002116    134  26.211679
102  9002116    135  30.067164
103  9002116    136  29.574803
104  9002116    137        NaN
105  9002116    138        NaN
106  9002116    139        NaN
107  9002116    140        NaN

* Slice 33 to 140.

In [57]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 19826 entries, 0 to 19825
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   case      19826 non-null  object 
 1   slice     19826 non-null  int64  
 2   distance  15601 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 619.6+ KB


* 4225 NaN

In [58]:
df = df.dropna(subset=["distance"]).reset_index(drop=True)

In [59]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15601 entries, 0 to 15600
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   case      15601 non-null  object 
 1   slice     15601 non-null  int64  
 2   distance  15601 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 365.8+ KB


In [60]:
slice_count_df = df.groupby('case').count()['slice'].reset_index()
slice_count_df = slice_count_df.sort_values(by='slice').reset_index(drop=True)

In [61]:
slice_count_df.head()

case  slice
0  9645478     37
1  9393987     48
2  9504862     50
3  9917284     51
4  9441420     51

In [62]:
slice_count_df.tail()

case  slice
191  9719999    107
192  9997856    108
193  9341240    109
194  9925760    113
195  9291078    124

In [63]:
slice_count_df.shape

(196, 2)

In [64]:
slice_count_df.describe()

slice
count  196.000000
mean    79.596939
std     15.319362
min     37.000000
25%     68.000000
50%     78.500000
75%     91.250000
max    124.000000

# IMPORTANT

* Some case only has 37 valid slices, and others up to 124 slices.

* To predict knee osteoarthritis (OA) severity/KL grades, we need the same feature vector for each case/patient to have the same length/size.

* Obstacle: some case has 37 valid slices with computed distance between femur and tibia and some has up to 124. To continue, we have to select 37 slices for each case for machine learning.

* NOTE: will come back to improve the code for problem 1 and 2 to increase valid slice for better prediction.

<hr>

In [65]:
new_df = df.groupby("case").head(37).reset_index(drop=True)

In [66]:
new_df.head(37)

case  slice   distance
0   9002116     33  12.754902
1   9002116     34  12.745283
2   9002116     35  13.036364
3   9002116     36  13.434783
4   9002116     37  13.372881
5   9002116     38  13.341667
6   9002116     39  13.090164
7   9002116     40  12.975806
8   9002116     41  13.070312
9   9002116     42  13.213740
10  9002116     43  13.146154
11  9002116     44  13.083969
12  9002116     45  12.932331
13  9002116     46  13.118519
14  9002116     47  13.389706
15  9002116     48  13.877698
16  9002116     49  14.433566
17  9002116     50  14.972414
18  9002116     51  15.074830
19  9002116     52  15.229730
20  9002116     53  15.553333
21  9002116     54  15.921569
22  9002116     55  16.311688
23  9002116     56  16.167742
24  9002116     57  16.547771
25  9002116     58  16.905660
26  9002116     59  17.525000
27  9002116     60  17.616352
28  9002116     61  17.394904
29  9002116     62  17.583333
30  9002116     63  18.382166
31  9002116     64  18.923077
32  9002116     65  19.743590
33  9002116     66  20.655844
34  9002116     67  21.228758
35  9002116     68  21.947368
36  9002116     69  23.420000

In [67]:
new_df.tail(37)

case  slice   distance
7215  9997856     28  16.746988
7216  9997856     29  17.375000
7217  9997856     30  18.297872
7218  9997856     31  18.696970
7219  9997856     32  17.500000
7220  9997856     33  17.429907
7221  9997856     34  17.189189
7222  9997856     35  17.226087
7223  9997856     36  17.302521
7224  9997856     37  17.147541
7225  9997856     38  17.388889
7226  9997856     39  17.061538
7227  9997856     40  16.857143
7228  9997856     41  16.941176
7229  9997856     42  16.942857
7230  9997856     43  16.542254
7231  9997856     44  16.455172
7232  9997856     45  16.197279
7233  9997856     46  16.306667
7234  9997856     47  16.274510
7235  9997856     48  16.141935
7236  9997856     49  16.108280
7237  9997856     50  16.275000
7238  9997856     51  15.801242
7239  9997856     52  15.635802
7240  9997856     53  15.456790
7241  9997856     54  15.581818
7242  9997856     55  15.432927
7243  9997856     56  15.716867
7244  9997856     57  17.187135
7245  9997856     58  18.097701
7246  9997856     59  18.202312
7247  9997856     60  18.802326
7248  9997856     61  24.571429
7249  9997856     62  23.660000
7250  9997856     63  24.891892
7251  9997856     64  27.571429

In [68]:
final_df = (
    new_df
    .groupby("case")["distance"]
    .apply(list)
    .reset_index(name="distance_vector")
)

In [69]:
final_df[final_df['case'] == '9002116']

case  \
0  9002116   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         distance_vector  
0  [12.754901960784316, 12.745283018867925, 13.036363636363635, 13.434782608695652, 13.372881355932204, 13.341666666666669, 13.09016393442623, 12.975806451612904, 13.0703125, 13.213740458015268, 13.146153846153846, 13.083969465648854, 12.93233082706767, 13.11851851851852, 13.389705882352942, 13.87769784172662, 14.433566433566432, 14.972413793103447, 15.07482993197279, 15.22972972972973, 15.553333333333333, 15.92156862745098, 16.31168831168831, 16.16774193548387, 16.547770700636942, 16.90566037735849, 17.525, 17.61635220125786, 17.394904458598727, 17.583333333333332, 18.38216560509554, 18.923076923076923, 19.743589743589745, 20.655844155844157, 21.22875816993464, 21.94736842105263, 23.42]

In [70]:
max_len = final_df["distance_vector"].str.len().max()
cols = [f"d{i}" for i in range(1, max_len+1)]

In [71]:
final_df[cols] = pd.DataFrame(final_df["distance_vector"].tolist(), index=final_df.index)
final_df = final_df.drop("distance_vector", axis=1)

In [72]:
data.head()

cases  V00XRKL V03KL
0  9002116        3     B
1  9005075        0     A
2  9005132        4     B
3  9026934        2     A
4  9030718        3     B

In [110]:
data = data.rename(columns={"cases": "case"})

In [113]:
print(data['case'].dtype)
print(type(data['case'][0]))

object
<class 'str'>


In [73]:
final_df.head()

case         d1         d2         d3         d4         d5         d6  \
0  9002116  12.754902  12.745283  13.036364  13.434783  13.372881  13.341667   
1  9005075  31.630952  31.597826  29.600000  27.888889  26.766990  25.409524   
2  9005132  12.602740  11.950617  11.379310  11.054348  11.212121  11.603774   
3  9026934  14.588235  14.337079  14.852632  15.080808  15.529412  16.103774   
4  9030718  31.166667  30.410000  28.771429  27.907407  26.609091  25.027273   

          d7         d8         d9        d10        d11        d12  \
0  13.090164  12.975806  13.070312  13.213740  13.146154  13.083969   
1  24.179245  23.289720  23.146789  22.554545  22.522124  22.184211   
2  11.594595  11.443478  11.697479  12.008130  12.000000  11.740157   
3  16.834862  16.630631  16.530973  15.964602  16.060345  16.588235   
4  24.892857  24.535714  28.338710  28.515873  28.352000  29.063492   

         d13        d14        d15        d16        d17        d18  \
0  12.932331  13.118519  13.389706  13.877698  14.433566  14.972414   
1  21.913043  21.530435  23.117647  22.663866  22.050847  22.403361   
2  11.635659  11.610687  11.753731  11.576642  11.340580  11.079137   
3  16.428571  16.516667  16.650000  17.256198  17.355372  17.942149   
4  29.080000  28.157025  24.513043  24.956897  25.120690  24.860870   

         d19        d20        d21        d22        d23        d24  \
0  15.074830  15.229730  15.553333  15.921569  16.311688  16.167742   
1  22.766667  23.557377  23.747967  24.370968  25.008000  27.148438   
2  10.741007  10.870504  10.492754  10.771429  10.637037  10.598540   
3  19.219512  19.458333  20.616667  21.864407  23.586207  28.247788   
4  25.034783  24.728070  25.513043  25.434783  26.724138  29.806723   

         d25        d26        d27        d28        d29        d30  \
0  16.547771  16.905660  17.525000  17.616352  17.394904  17.583333   
1  29.389313  29.887218  32.569343  33.394161  34.316547  35.964539   
2  11.086957  11.704225  11.500000  12.171642  12.740741  13.598485   
3  31.672727  32.788462  22.392857  22.722892  23.764228  23.943548   
4  29.728814  30.529412  30.415254  29.170940  30.752137  31.465517   

         d31        d32        d33        d34        d35        d36        d37  
0  18.382166  18.923077  19.743590  20.655844  21.228758  21.947368  23.420000  
1  36.773050  36.237410  38.524823  38.414286  38.690647  39.597122  40.304348  
2  14.115385  15.955224  17.777778  17.800000  17.896296  18.014925  16.230088  
3  23.209677  24.226562  23.750000  23.480620  23.700000  22.131783  23.427481  
4  31.543103  31.061404  31.327434  22.785714  19.689655  17.989247  26.477876

In [82]:
print(final_df['case'].dtype)
print(type(final_df['case'][0]))

object
<class 'str'>


In [114]:
# Merge:

final_df_merged = final_df.merge(data, on="case", how="left")

In [116]:
final_df_merged = final_df_merged.rename(columns={
    'V00XRKL':'y1', 
    'V03KL':'y2'})

In [117]:
final_df_merged.head()

case         d1         d2         d3         d4         d5         d6  \
0  9002116  12.754902  12.745283  13.036364  13.434783  13.372881  13.341667   
1  9005075  31.630952  31.597826  29.600000  27.888889  26.766990  25.409524   
2  9005132  12.602740  11.950617  11.379310  11.054348  11.212121  11.603774   
3  9026934  14.588235  14.337079  14.852632  15.080808  15.529412  16.103774   
4  9030718  31.166667  30.410000  28.771429  27.907407  26.609091  25.027273   

          d7         d8         d9        d10        d11        d12  \
0  13.090164  12.975806  13.070312  13.213740  13.146154  13.083969   
1  24.179245  23.289720  23.146789  22.554545  22.522124  22.184211   
2  11.594595  11.443478  11.697479  12.008130  12.000000  11.740157   
3  16.834862  16.630631  16.530973  15.964602  16.060345  16.588235   
4  24.892857  24.535714  28.338710  28.515873  28.352000  29.063492   

         d13        d14        d15        d16        d17        d18  \
0  12.932331  13.118519  13.389706  13.877698  14.433566  14.972414   
1  21.913043  21.530435  23.117647  22.663866  22.050847  22.403361   
2  11.635659  11.610687  11.753731  11.576642  11.340580  11.079137   
3  16.428571  16.516667  16.650000  17.256198  17.355372  17.942149   
4  29.080000  28.157025  24.513043  24.956897  25.120690  24.860870   

         d19        d20        d21        d22        d23        d24  \
0  15.074830  15.229730  15.553333  15.921569  16.311688  16.167742   
1  22.766667  23.557377  23.747967  24.370968  25.008000  27.148438   
2  10.741007  10.870504  10.492754  10.771429  10.637037  10.598540   
3  19.219512  19.458333  20.616667  21.864407  23.586207  28.247788   
4  25.034783  24.728070  25.513043  25.434783  26.724138  29.806723   

         d25        d26        d27        d28        d29        d30  \
0  16.547771  16.905660  17.525000  17.616352  17.394904  17.583333   
1  29.389313  29.887218  32.569343  33.394161  34.316547  35.964539   
2  11.086957  11.704225  11.500000  12.171642  12.740741  13.598485   
3  31.672727  32.788462  22.392857  22.722892  23.764228  23.943548   
4  29.728814  30.529412  30.415254  29.170940  30.752137  31.465517   

         d31        d32        d33        d34        d35        d36  \
0  18.382166  18.923077  19.743590  20.655844  21.228758  21.947368   
1  36.773050  36.237410  38.524823  38.414286  38.690647  39.597122   
2  14.115385  15.955224  17.777778  17.800000  17.896296  18.014925   
3  23.209677  24.226562  23.750000  23.480620  23.700000  22.131783   
4  31.543103  31.061404  31.327434  22.785714  19.689655  17.989247   

         d37  y1 y2  
0  23.420000   3  B  
1  40.304348   0  A  
2  16.230088   4  B  
3  23.427481   2  A  
4  26.477876   3  B

<hr>

# Write final_df ready for meachine learning to 'final_df.csv'. Go to problem 4 to continue:

In [119]:
final_df_merged.to_csv("../final_problem4-5-6_kien_steven/final_df_merged.csv", index=False)